Adding network statistics to container metrics

shashidharatd opened the pull request:
Network statistics for containers is read from sysfs. Now we can get data such as no.of.bytes, packets, error and dropped packets both recieved and transferred from the container. The code is picked from docker's libcontainer.

I did also check the performance impact and found that the implementation takes very little time to collect the network statistics.

glyn commented:
I've made a couple of inline comments about returning zero values along with errors and testing the network stats.

On the libcontainer front, please could you say which sections of code were copied from libcontainer? Also, IANAL but you don't seem to have followed section 4.c of the [Apache v2 license](http://www.apache.org/licenses/LICENSE-2.0.html) in respect of the libcontainer code. I'm thinking specifically of copyright and attribution notice.

shashidharatd commented:
HI @glyn,
Thanks for the review, i would address the review comments.
Actually the code is not from libcontainer, but docker daemon's execdriver. the function i picked is `readSysfsNetworkStats`. Here is the link https://github.com/docker/docker/blob/master/daemon/execdriver/driver_linux.go#L66

Its a simple technique of reading from sysfs. we could write our own function to do the job. But definitely the idea came from comparative study of container metrics. On the other hand i didn't want to import the complete package of execdriver, just for some utility function.

Now i am clueless on how to go about it with the licensing issue. so request to take a call on this matter. please feel free to reject this PR, if it violates some license.

Thanks
@shashidharatd

glyn commented:
You could avoid the IP issue by rewriting the function from scratch.

shashidharatd commented:
ok, i will write the utility function from scratch.

glyn commented:
Thanks! If you can develop it with its own test, that would be great.

shashidharatd commented:
Hi @glyn,
   Sorry for the delayed response. Writing the tests really took some time.
Now I have rewritten the code and added the code for testing metrics for network device.
I have used the ping utility to test number of packets sent and received. However getting accurate number of packets is difficult because there are ARP messages sent and received on the interface. I have set static ARP entries and disabled IPv6, but still some IPv6 ND RA transmissions are received
while testing. We cannot control the incoming packets, So i have used a delay of 3 seconds after the link is created and then tested the number of packets sent and received using ping utility.

Still had not tried to test error packets and drop packets statistics. Wanted to know if the testing method was OK or is there a better way to test? please suggest.

Thanks
@shashidharatd

glyn commented:
That sounds like a reasonable way of doing integration testing. We need to avoid tests which will fail randomly, so it may be worth sending a 10 or so pings and then ensuring the number of packets is within a certain range.

You might consider adding a unit test where you inject a fake for accessing the network stats file. Then you can check the rest of the logic is ok. However this may not be worth it, depending on the complexity of that logic.

shashidharatd commented:
I would like to keep it generic and include only RxBytes and TxBytes and remove other stats which would be useful to implement in other backends. 

As suggested i would test whether the number of bytes transferred and received are within certain range to account for other packets entering and exiting the interface.

glyn commented:
Sounds good. (Presumably you meant "... remove other stats which would be difficult to implement in other backends".)

shashidharatd commented:
yes :) Sorry for the mess up in the sentence.

glyn commented:
No problem - just checking to be on the safe side.

julz commented on old/main.go:
```diff
@@ -228,6 +228,16 @@ func Main() {
 	}
 
+	// sysfs is required for reading network statistics
+	mntpoint, _ := os.Stat("/sys")
```
we should check the errors here

julz commented on old/main.go:
```diff
@@ -228,6 +228,16 @@ func Main() {
+	mntpoint, _ := os.Stat("/sys")
+	parent, _ := os.Stat("/")
+
+	if mntpoint.Sys().(*syscall.Stat_t).Dev == parent.Sys().(*syscall.Stat_t).Dev {
```
I think it'd probably be better if this logic went in to a utility function (even better if it could be tested); we normally try to avoid logic in main.go. Maybe linux_backend.Start() would be an appropriate place for something like this?

julz commented on network/devices/link_linux_test.go:
```diff
@@ -166,4 +166,73 @@ var _ = Describe("Link Management", func() {
+	Describe("GetStatistics", func() {
+		BeforeEach(func() {
+			cmd, _ := gexec.Start(exec.Command("sh", "-c", "mount -n -t sysfs sysfs /sys ||true"), GinkgoWriter, GinkgoWriter)
+			Expect(cmd.Wait("2s").ExitCode()).To(Equal(0))
```
this is a very minor thing, but you could use [`Eventually(cmd).Should(gexec.Exit(0))`](http://onsi.github.io/gomega/#gexec-testing-external-processes) here

julz commented on network/devices/link_linux_test.go:
```diff
@@ -166,4 +166,73 @@ var _ = Describe("Link Management", func() {
+					`,
+				), GinkgoWriter, GinkgoWriter)
+				Expect(err).ToNot(HaveOccurred())
+				Expect(cmd.Wait("10s").ExitCode()).To(Equal(0))
```
see comment above about `gexec.Exit()`

julz commented on network/devices/link.go:
```diff
@@ -78,6 +99,33 @@ func (Link) List() (names []string, err error) {
+		sysFile string
+	}
+
+	netStats := []netStatsPair{
```
rather than looping over this array, why not just have a `stat(name string)` function and then this method would become :

~~~~
return garden.ContainerNetworkStat{
  RxBytes: stat("rx_bytes"),
  TxBytes: stat("tx_bytes"),
}
~~~~

I had to spend a fair while reading this method to figure out what it did, I think the above would be quite a lot easier to quickly understand

julz commented on network/devices/link.go:
```diff
@@ -78,6 +99,33 @@ func (Link) List() (names []string, err error) {
 	return names, nil
 }
 
+func (l Link) GetStatistics() (garden.ContainerNetworkStat, error) {
```
it's probably slightly more idiomatic Go to have the method be called `Statistics` rather than `GetStatistics`

julz commented on linux_container/metrics_test.go:
```diff
@@ -241,5 +245,45 @@ system 2
+						new(fake_process_tracker.FakeProcessTracker),
+						process.Env{"env1": "env1Value", "env2": "env2Value"},
+						new(networkFakes.FakeFilter),
+						devices.NewLink("non-existent"),
```
this should probably use a fake rather than a real link from the device package, container tests should run on any system (so dependencies are normally mocked out) whereas stuff in the devices package is system-specific.

julz commented on container_pool/container_pool.go:
```diff
@@ -278,6 +279,7 @@ func (p *LinuxContainerPool) Create(spec garden.ContainerSpec) (c linux_backend.
 		process_tracker.New(containerPath, p.runner),
 		rootFSEnv.Merge(specEnv),
 		p.filterProvider.ProvideFilter(id),
+		devices.NewLink(p.sysconfig.NetworkInterfacePrefix+id+"-0"),
```
wonder if we could find a way to test this logic (and avoid duplicating it)

julz commented on linux_container/linux_container.go:
```diff
@@ -93,6 +94,8 @@ type LinuxContainer struct {
 
 	processIDPool *ProcessIDPool
+
+	link devices.LinkInterface
```
rather than having linux_container depend on the whole devices.LinkInterface (most of which it doesn't need), perhaps it could depend just on a `type NetworkStatisticser interface { Statistics() }` defined locally in this package?

shashidharatd commented:
Thanks for your time and comments @julz will try to address the comments

shashidharatd commented:
Hi @julz,
    I have reworked the code for your comments. I hope all the changes are according to your expectation. please check the fixes.

Thank you
@shashidharatd

craigfurman commented:
Merged.

Thanks @shashidharatd!

@craigfurman and @julz

shashidharatd commented:
Thank you @julz  & @craigfurman
